### PR TITLE
Decouple HTTP Servers from cli/commands/daemon

### PR DESCRIPTION
**Pull request by lidel**
> Part of an effort to run embedded js-ipfs in Brave :lion: https://github.com/ipfs-shipyard/ipfs-companion/issues/664  

### Summary

This PR decouples HTTP servers from code `jsipfs daemon`.
Apart from cleanup, we need this to enable HTTP server in browser context with [`chrome.sockets`](https://github.com/ipfs-shipyard/ipfs-companion/issues/664).

**There are no functional changes.** I just moved non-HTTP code  to `src/cli/standalone-daemon.js`

### Background

In the past API was exposed via HTTP Server only when jsipfs daemon was run from the commandline, so src/http/index.js was also responsible for orchestration that is not related to HTTP itself. The coupling was quite bad: `HttpApi` basically decided on libp2p configuration: it enabled WebRTC if available, TCP transport, MulticastDNS etc.

### Changes

This refactor moves code that is not related to HTTP Servers into `standalone-daemon.js`, which is easier to reason about, and unlocks use of `HttpApi` in contexts other than CLI  `jsipfs daemon`, such as Firefox with libdweb or [Chromium-based web browser with `chrome.sockets` APIs](https://github.com/ipfs-shipyard/ipfs-companion/issues/664)

With this PR, HttpApi is not only decoupled from Standalone Daemon, but  can also be manually started/stopped in browser context.

I believe only ipfs-companion will be using this for now, but it is a fairly generic code.
Someone else should be able to use this in Chrome App environment:

#### Start
```js
const Ipfs = require('ipfs')
const HttpApi = require('ipfs/src/http')

let node = new Ipfs(ipfsOpts)
let nodeHttpApi

node.on('start', async () => {
  // HttpApi is off in browser context and needs to be started separately
  const httpServers = new HttpApi(node, ipfsOpts)
  nodeHttpApi = await httpServers.start()
})
```
#### Stop
```js
if (nodeHttpApi) {
  await nodeHttpApi.stop()
  nodeHttpApi = null
}
if (node) {
  await node.stop()
  node = null
}
```

### Checklist

- [x] do the thing ;-)
- [x] fix printing of API and Gateway port in browser console
     - DONE: [when run in browser](https://github.com/ipfs-shipyard/ipfs-companion/issues/664#issuecomment-481697875), it will announce using `debug`:
       > 

- [x] ensure issue with returning payload in browser context is not due to js-ipfs
      - DONE: does not block this PR (it was due to Node-specific check in Hapi, will be addressed upstream)

Refs.
https://github.com/ipfs-shipyard/ipfs-companion/issues/664, https://github.com/ipfs/js-ipfs/issues/1855

### CODE REVIEW

**Comment by lidel** [comment]
(WILL REBASE SHORTLY)

**Comment by lidel** [comment]
_Ok, should be ready, let's review it._

File: src/http/index.js [hunk 1]
```diff
@@ -119,14 +77,15 @@ class HttpApi {
...
-      ipfs._print('API listening on %s', apiServer.info.ma)
+      announce('API listening on %s', apiServer.info.ma.toString())
     })
     this._gatewayServers.forEach(gatewayServer => {
-      ipfs._print('Gateway (read only) listening on %s', gatewayServer.info.ma)
+      announce('Gateway (read only) listening on %s', gatewayServer.info.ma.toString())
     })
     this._apiServers.forEach(apiServer => {
-      ipfs._print('Web UI available at %s', toUri(apiServer.info.ma) + '/webui')
+      announce('Web UI available at %s', toUri(apiServer.info.ma) + '/webui')
```
Comment (alanshaw) [inline src/http/index.js]: Shall we move this into `src/cli/commands/daemon.js`?

File: src/cli/commands/daemon.js [hunk 2]
```diff
@@ -44,8 +44,8 @@ module.exports = {
       const repoPath = getRepoPath()
 
       // Required inline to reduce startup time
-      const HttpApi = require('../../http')
-      const api = new HttpApi({
+      const StandaloneDaemon = require('../../cli/standalone-daemon')
```
Comment (alanshaw) [inline src/cli/commands/daemon.js]: Can we just call this `Daemon`?

File: src/cli/standalone-daemon.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    const ipfs = new IPFS(ipfsOpts)
+
+    await new Promise((resolve, reject) => {
+      ipfs.once('error', err => {
+        this._log('error starting core', err)
+        err.code = 'ENOENT'
+        reject(err)
+      })
+      ipfs.once('start', resolve)
+    })
```
Comment (alanshaw) [inline src/cli/standalone-daemon.js]: Argh, any chance you can help @pkafei writing a test for https://github.com/ipfs/js-ipfs/pull/1878 so we don't need to do this?

(completely optional and I won't block merging this!)

Comment (pkafei) [reply src/cli/standalone-daemon.js]: Can we pair program and work on the test together?

Comment (lidel) [reply src/cli/standalone-daemon.js]: @alanshaw I'd prefer to merge this PR as-is and improve it later.

@pkafei Sure! I am booked today, but feel free to grab time slot next week: https://calendly.com/pl-lidel/ :)

File: src/cli/commands/daemon.js (hunk 2 above)
Comment (lidel) [reply src/cli/commands/daemon.js]: Done in https://github.com/ipfs/js-ipfs/pull/1950/commits/a2a307d3cf2887ea5353585e5ac1b8a17eef201f

File: src/http/index.js (hunk 1 above)
Comment (lidel) [reply src/http/index.js]: Good idea! Moved in https://github.com/ipfs/js-ipfs/pull/1950/commits/a2a307d3cf2887ea5353585e5ac1b8a17eef201f

File: src/cli/daemon.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    this._httpApi._apiServers.forEach(apiServer => {
+      ipfs._print('API listening on %s', apiServer.info.ma.toString())
+    })
+    this._httpApi._gatewayServers.forEach(gatewayServer => {
+      ipfs._print('Gateway (read only) listening on %s', gatewayServer.info.ma.toString())
+    })
+    this._httpApi._apiServers.forEach(apiServer => {
+      ipfs._print('Web UI available at %s', toUri(apiServer.info.ma) + '/webui')
+    })
```
Comment (alanshaw) [inline src/cli/daemon.js]: I meant move to ` src/cli/commands/daemon.js` - where the rest of the logging already is.

Comment (lidel) [reply src/cli/daemon.js]: Ack, done in https://github.com/ipfs/js-ipfs/pull/1950/commits/1e821029b75d5e5847cdd26007631e0bc6861477

File: src/cli/commands/daemon.js [hunk 5]
```diff
@@ -60,7 +61,16 @@ module.exports = {
...
+        await daemon.start()
+        daemon._httpApi._apiServers.forEach(apiServer => {
+          print(`API listening on ${apiServer.info.ma.toString()}`)
+        })
+        daemon._httpApi._gatewayServers.forEach(gatewayServer => {
+          print(`Gateway (read only) listening on ${gatewayServer.info.ma.toString()}`)
+        })
+        daemon._httpApi._apiServers.forEach(apiServer => {
+          print(`Web UI available at ${toUri(apiServer.info.ma)}/webui`)
+        })
```
Comment (alanshaw) [inline src/cli/commands/daemon.js]: 👌

**Comment by alanshaw** [comment]
Thanks @lidel ❤️

**Pull request merged**